File: Yandex/Practicum/Algorightms_and_Data_Structures/8_sprint/C._Самый_длинный_палиндром_2/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(M) -> O(1)
    """
    sequence: str = sys.stdin.readline().rstrip()
    histogram: dict[int, int] = dict()
    for letter in sequence:
        histogram[ord(letter)] = histogram.get(ord(letter), 0) + 1
    part: list[str] = []
    median: str = ''
    for idx in range(97, 123):
        if idx not in histogram: continue
        for _ in range(histogram[idx] // 2): part.append(chr(idx))
        if histogram[idx] & 1 and not median: median = chr(idx)

    sys.stdout.write('{}{}{}\n'.format(''.join(part), median, ''.join(reversed(part))))


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/8_sprint/C._Самый_длинный_палиндром_2/solution.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <algorithm>


void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(M) => O(1)
    */
    std::string sequence;
    std::cin >> sequence;
    std::unordered_map<int, int> histogram;
    for (char& letter : sequence) ++histogram[(int)letter];
    char median = '0';
    std::string part = "";
    for (int idx = 97; idx <= 122; ++idx) {
        if (!histogram.count(idx)) continue;
        for (int i = 0; i < histogram[idx] / 2; ++i) part.push_back((char)idx);
        if (histogram[idx] % 2 != 0 && median == '0') median = (char)idx;
    }
    std::string palindrome = part;
    if (median != '0') palindrome.push_back(median);
    std::reverse(part.begin(), part.end());
    palindrome += part;
    std::cout << palindrome << "\n";
}


int main() {
    solution();
}